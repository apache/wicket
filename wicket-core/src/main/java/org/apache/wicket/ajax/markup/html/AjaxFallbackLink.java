/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.wicket.ajax.markup.html;

import org.apache.wicket.ajax.AbstractDefaultAjaxBehavior;
import org.apache.wicket.ajax.AjaxChannel;
import org.apache.wicket.ajax.AjaxEventBehavior;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.ajax.attributes.AjaxRequestAttributes;
import org.apache.wicket.markup.ComponentTag;
import org.apache.wicket.markup.html.link.Link;
import org.apache.wicket.model.IModel;

/**
 * An ajax link that will degrade to a normal request if ajax is not available or javascript is
 * disabled.
 * <p>
 * If JavaScript is enabled then the registered JavaScript event 'click' handler will be used,
 * otherwise the 'href' attribute if the markup element is an &lt;a&gt;, &lt;area&gt; or &lt;link&gt;.
 * AjaxFallbackLink doesn't fallback if the markup element is none of the three above.
 * </p>
 *
 * @since 1.2
 * 
 * @author Igor Vaynberg (ivaynberg)
 * @param <T>
 *            type of model object
 */
public abstract class AjaxFallbackLink<T> extends Link<T> implements IAjaxLink
{
	/** */
	private static final long serialVersionUID = 1L;

	/**
	 * Construct.
	 * 
	 * @param id
	 */
	public AjaxFallbackLink(final String id)
	{
		this(id, null);
	}

	/**
	 * Construct.
	 * 
	 * @param id
	 * @param model
	 */
	public AjaxFallbackLink(final String id, final IModel<T> model)
	{
		super(id, model);

	}

	@Override
	protected void onInitialize()
	{
		super.onInitialize();
		add(newAjaxEventBehavior("click"));
	}

	/**
	 * @param event
	 *            the name of the default event on which this link will listen to
	 * @return the ajax behavior which will be executed when the user clicks the link
	 */
	protected AjaxEventBehavior newAjaxEventBehavior(String event)
	{
		return new AjaxEventBehavior(event)
		{
			private static final long serialVersionUID = 1L;

			@Override
			protected void onEvent(AjaxRequestTarget target)
			{
				onClick(target);
			}

			/**
			 * 
			 * @see org.apache.wicket.ajax.AjaxEventBehavior#onComponentTag(org.apache.wicket.markup.ComponentTag)
			 */
			@Override
			protected void onComponentTag(ComponentTag tag)
			{
				// only render handler if link is enabled
				if (isLinkEnabled())
				{
					super.onComponentTag(tag);
				}
			}

			@Override
			protected AjaxChannel getChannel()
			{
				return AjaxFallbackLink.this.getChannel();
			}

			@Override
			protected void updateAjaxAttributes(AjaxRequestAttributes attributes)
			{
				super.updateAjaxAttributes(attributes);
				AjaxFallbackLink.this.updateAjaxAttributes(attributes);
			}
		};
	}

	/**
	 * @param attributes
	 */
	protected void updateAjaxAttributes(AjaxRequestAttributes attributes)
	{
	}

	/**
	 * @return the channel that manages how Ajax calls are executed
	 * @see AbstractDefaultAjaxBehavior#getChannel()
	 */
	@Deprecated
	protected AjaxChannel getChannel()
	{
		return null;
	}

	/**
	 * 
	 * @see org.apache.wicket.markup.html.link.Link#onClick()
	 */
	@Override
	public final void onClick()
	{
		onClick(null);
	}

	/**
	 * Callback for the onClick event. If ajax failed and this event was generated via a normal link
	 * the target argument will be null
	 * 
	 * @param target
	 *            ajax target if this linked was invoked using ajax, null otherwise
	 */
	@Override
	public abstract void onClick(final AjaxRequestTarget target);

	/**
	 * Removes any inline 'onclick' attributes set by Link#onComponentTag(ComponentTag).
	 *
	 * @param tag
	 *            the component tag
	 */
	@Override
	protected void onComponentTag(ComponentTag tag)
	{
		super.onComponentTag(tag);

		// Ajax links work with JavaScript Event registration
		tag.remove("onclick");
	}


}
