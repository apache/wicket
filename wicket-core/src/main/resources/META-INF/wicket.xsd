<?xml version="1.0" encoding="UTF-8"?>
<schema xmlns="http://www.w3.org/2001/XMLSchema" targetNamespace="http://wicket.apache.org"
	xmlns:wicket="http://wicket.apache.org"
	xmlns:xhtml="http://www.w3.org/1999/xhtml"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://www.w3.org/1999/xhtml http://www.w3.org/MarkUp/SCHEMA/xhtml-basic10.xsd"
	elementFormDefault="qualified"
	version="7.0.0">
	
	<redefine schemaLocation="http://www.w3.org/MarkUp/SCHEMA/xhtml-basic10.xsd">
		<attributeGroup name="xhtml.label.attlist">
			<attributeGroup ref="xhtml.label.attlist"/>
			<attribute ref="wicket:for"/>
		</attributeGroup>
	</redefine>

	<!-- types -->

	<simpleType name="token">
		<annotation>
			<documentation>An extension of standard token type with the
				restriction of collapsing white space, which means that the XML
				processor WILL REMOVE all white space characters (line feeds, tabs,
				spaces, carriage returns are replaced with spaces, leading and
				trailing spaces are removed, and multiple spaces are reduced to a
				single space).</documentation>
		</annotation>
		<restriction base="token">
			<whiteSpace value="collapse" />
		</restriction>
	</simpleType>

	<simpleType name="notEmptyToken">
		<annotation>
			<documentation>Extends wicket:token type with the restriction that
				the value can not be empty.</documentation>
		</annotation>
		<restriction base="wicket:token">
			<minLength value="1" />
		</restriction>
	</simpleType>

	<complexType name="mixedContent" mixed="true">
		<annotation>
			<documentation>A complex type that allows any amount of mixed child
				content from any origin.</documentation>
		</annotation>
		<choice minOccurs="0" maxOccurs="unbounded">
			<any namespace="##targetNamespace" processContents="strict" />
			<any namespace="##other" processContents="lax" />
		</choice>
	</complexType>

	<complexType name="notEmptyMixedContent" mixed="true">
		<annotation>
			<documentation>Extends wicket:mixedContent with the restriction that
				at least one child element must be given.</documentation>
		</annotation>
		<choice minOccurs="1" maxOccurs="unbounded">
			<any namespace="##targetNamespace" processContents="strict" />
			<any namespace="##other" processContents="lax" />
		</choice>
	</complexType>

	<!-- attributes -->

	<attribute name="enclosure">
		<annotation>
			<documentation>This attribute is useful for filtering markup that
				surrounds a component but has its visibility dependent on the
				visibility of that component. It is similar to the wicket:enclosure
				tag. The value is the wicket:id of the dependent component. No value
				can be used if there is only one nested wicket component.
				"foo:bar:baz" format can be used to specify a nested path.
			</documentation>
		</annotation>
		<simpleType>
			<restriction base="wicket:notEmptyToken">
				<pattern value="([^:]+(:[^:]+)*)?" />
			</restriction>
		</simpleType>
	</attribute>

	<attribute name="id" type="wicket:notEmptyToken">
		<annotation>
			<documentation>The wicket component id. Used on any element to which
				you want to add a component.</documentation>
		</annotation>
	</attribute>

	<attribute name="for">
		<annotation>
			<documentation>Can be used in HTML <![CDATA[<label]]> element to automatically link it
				to a form element. The value of this attribute must have the same value of
				the form component's wicket:id attribute.
			</documentation>
		</annotation>
		<simpleType>
			<restriction base="wicket:notEmptyToken">
				<pattern value="([^:]+(:[^:]+)*)?" />
			</restriction>
		</simpleType>
	</attribute>

	<attribute name="message">
		<annotation>
			<documentation>Used on any tag that we want Wicket to provide an
				attribute with a value that's the result of a resource lookup.
				Format: "att1:key1,att2:key2", where att1, att2 are attribute names
				and key1, key2 are resource id-s.
			</documentation>
		</annotation>
		<simpleType>
			<restriction base="wicket:notEmptyToken">
				<pattern value="[^:,]+:[^:,]+(,[^:,]+:[^:,]+)*" />
			</restriction>
		</simpleType>
	</attribute>

	<!-- elements -->

	<element name="body">
		<annotation>
			<documentation>Used with wicket:border to specify the location of the
				original content.</documentation>
		</annotation>
		<complexType />
	</element>

	<element name="border" type="wicket:notEmptyMixedContent">
		<annotation>
			<documentation>Specifies bordering content for a wicket:body.
			</documentation>
		</annotation>
		<unique name="borderWicketId">
			<selector xpath="./*" />
			<field xpath="@wicket:id" />
		</unique>
	</element>

	<element name="child">
		<annotation>
			<documentation>Wicket will replace this content with the markup of
				the derived component (see wicket:extend).</documentation>
		</annotation>
		<complexType />
	</element>

	<element name="header-items">
		<annotation>
			<documentation><![CDATA[A special element that is used as a placeholder inside the HTML <head>
			element. Wicket will replace it with all header contributions.]]></documentation>
		</annotation>
		<complexType />
	</element>

	<element name="container">
		<annotation>
			<documentation>A tag to which you can attach a wicket component but
				it will not produce a html tag in the result.</documentation>
		</annotation>
		<complexType mixed="true">
			<complexContent>
				<extension base="wicket:mixedContent">
					<attribute ref="wicket:id" use="required" />
				</extension>
			</complexContent>
		</complexType>
		<unique name="containerWicketId">
			<selector xpath="./*" />
			<field xpath="@wicket:id" />
		</unique>
	</element>

	<element name="enclosure">
		<annotation>
			<documentation>This tag is useful for filtering markup that surrounds
				a component but has its visibility dependent on the visibility of
				that component. It is similar to the wicket:enclosure attribute.
			</documentation>
		</annotation>
		<complexType mixed="true">
			<complexContent>
				<extension base="wicket:notEmptyMixedContent">
					<attribute name="child" use="optional">
						<annotation>
							<documentation>Specifies the wicket:id of the dependent
								component. This attribute can be omitted if there is only one
								nested wicket component. "foo:bar:baz" format can be used to
								specify a nested path.</documentation>
						</annotation>
						<simpleType>
							<restriction base="wicket:notEmptyToken">
								<pattern value="[^:]+(:[^:]+)*" />
							</restriction>
						</simpleType>
					</attribute>
				</extension>
			</complexContent>
		</complexType>
		<unique name="enclosureWicketId">
			<selector xpath="./*" />
			<field xpath="@wicket:id" />
		</unique>
	</element>

	<element name="extend" type="wicket:mixedContent">
		<annotation>
			<documentation>Extend the markup of the superclass with this content.
				Used together with wicket:child.</documentation>
		</annotation>
		<unique name="extendWicketId">
			<selector xpath="./*" />
			<field xpath="@wicket:id" />
		</unique>
	</element>

	<element name="fragment">
		<annotation>
			<documentation>Is similar to wicket:panel but it is declared in the
				parent's markup instead of in a separate markup file.
			</documentation>
		</annotation>
		<complexType mixed="true">
			<complexContent>
				<extension base="wicket:mixedContent">
					<attribute ref="wicket:id" use="required" />
				</extension>
			</complexContent>
		</complexType>
		<unique name="fragmentWicketId">
			<selector xpath="./*" />
			<field xpath="@wicket:id" />
		</unique>
	</element>

	<element name="head" type="wicket:mixedContent">
		<annotation>
			<documentation>Used for header contributions. Using this, panels can
				add header sections to the pages they are placed on.</documentation>
		</annotation>
		<unique name="headWicketId">
			<selector xpath="./*" />
			<field xpath="@wicket:id" />
		</unique>
	</element>

	<element name="link" type="wicket:mixedContent">
		<annotation>
			<documentation>Support for wicket's autolink functionality. Normally,
				you need to add a component (for example, a BookmarkablePageLink)
				for each link that Wicket manages. Using the wicket:link tag will
				automatically do this in the background for you.</documentation>
		</annotation>
		<unique name="linkWicketId">
			<selector xpath="./*" />
			<field xpath="@wicket:id" />
		</unique>
	</element>

	<element name="message">
		<annotation>
			<documentation>Wicket will replace this with a string that is
				retrieved from a resource bundle.</documentation>
		</annotation>
		<complexType mixed="true">
			<complexContent>
				<extension base="wicket:mixedContent">
					<attribute name="key" use="required" type="wicket:notEmptyToken">
						<annotation>
							<documentation>The resource id used for lookup.</documentation>
						</annotation>
					</attribute>
					<attribute name="escape" use="optional" type="boolean" default="false">
						<annotation>
							<documentation>Defines whether the resource value should be escaped or not.</documentation>
						</annotation>
					</attribute>
				</extension>
			</complexContent>
		</complexType>
		<unique name="messageWicketId">
			<selector xpath="./*" />
			<field xpath="@wicket:id" />
		</unique>
	</element>

	<element name="panel" type="wicket:mixedContent">
		<annotation>
			<documentation>The wicket:panel tag surrounds a component. This lets
				you wrap the component with HTML and BODY tags (so it can be
				displayed in a browser) but, when you include it, only the content
				inside the wicket:panel tag is actually included.</documentation>
		</annotation>
		<unique name="panelWicketId">
			<selector xpath="./*" />
			<field xpath="@wicket:id" />
		</unique>
	</element>

	<element name="remove" type="wicket:mixedContent">
		<annotation>
			<documentation>Wicket will remove this content in the final markup.
				This is useful for when you want your web designer to be able to
				show repeated content when they're working on it, but you want to
				generate that content using a ListView (or other loop).
			</documentation>
		</annotation>
	</element>
</schema>
