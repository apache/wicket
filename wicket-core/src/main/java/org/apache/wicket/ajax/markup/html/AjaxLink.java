/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.wicket.ajax.markup.html;

import org.apache.wicket.IGenericComponent;
import org.apache.wicket.ajax.AjaxEventBehavior;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.ajax.attributes.AjaxRequestAttributes;
import org.apache.wicket.markup.ComponentTag;
import org.apache.wicket.markup.html.link.AbstractLink;
import org.apache.wicket.model.IModel;

/**
 * A component that allows a trigger request to be triggered via html anchor tag
 * 
 * @since 1.2
 * 
 * @author Igor Vaynberg (ivaynberg)
 * @param <T>
 *            type of model object
 * 
 */
public abstract class AjaxLink<T> extends AbstractLink implements IAjaxLink, IGenericComponent<T>
{
	private static final long serialVersionUID = 1L;

	/**
	 * Construct.
	 * 
	 * @param id
	 */
	public AjaxLink(final String id)
	{
		this(id, null);
	}

	/**
	 * Construct.
	 * 
	 * @param id
	 * @param model
	 */
	public AjaxLink(final String id, final IModel<T> model)
	{
		super(id, model);
	}


	@Override
	protected void onInitialize()
	{
		super.onInitialize();
		add(newAjaxEventBehavior("onclick"));
	}

	/**
	 * @param event
	 *            the name of the default event on which this link will listen to
	 * @return the ajax behavior which will be executed when the user clicks the link
	 */
	protected AjaxEventBehavior newAjaxEventBehavior(String event)
	{
		return new AjaxEventBehavior(event)
		{
			private static final long serialVersionUID = 1L;

			@Override
			protected void onEvent(AjaxRequestTarget target)
			{
				onClick(target);
			}

			@Override
			protected void updateAjaxAttributes(AjaxRequestAttributes attributes)
			{
				super.updateAjaxAttributes(attributes);
				AjaxLink.this.updateAjaxAttributes(attributes);
			}
		};
	}

	protected void updateAjaxAttributes(AjaxRequestAttributes attributes)
	{
	}

	@Override
	protected void onComponentTag(ComponentTag tag)
	{
		super.onComponentTag(tag);

		if (isEnabledInHierarchy())
		{
			// disable any href attr in markup
			if (tag.getName().equalsIgnoreCase("a") || tag.getName().equalsIgnoreCase("link") ||
				tag.getName().equalsIgnoreCase("area"))
			{
				tag.put("href", "javascript:;");
			}
		}
		else
		{
			disableLink(tag);
		}

	}

	/**
	 * Listener method invoked on the ajax request generated when the user clicks the link
	 * 
	 * @param target
	 */
	@Override
	public abstract void onClick(final AjaxRequestTarget target);

	@Override
	@SuppressWarnings("unchecked")
	public final IModel<T> getModel()
	{
		return (IModel<T>)getDefaultModel();
	}

	@Override
	public final void setModel(IModel<T> model)
	{
		setDefaultModel(model);
	}

	@Override
	@SuppressWarnings("unchecked")
	public final T getModelObject()
	{
		return (T)getDefaultModelObject();
	}

	@Override
	public final void setModelObject(T object)
	{
		setDefaultModelObject(object);
	}

}
