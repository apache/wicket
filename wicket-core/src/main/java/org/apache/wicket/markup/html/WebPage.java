/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.wicket.markup.html;

import org.apache.wicket.Component;
import org.apache.wicket.Page;
import org.apache.wicket.markup.MarkupType;
import org.apache.wicket.markup.head.IHeaderResponse;
import org.apache.wicket.markup.html.internal.HtmlHeaderContainer;
import org.apache.wicket.markup.html.link.BookmarkablePageLink;
import org.apache.wicket.markup.parser.filter.HtmlHeaderSectionHandler;
import org.apache.wicket.markup.renderStrategy.AbstractHeaderRenderStrategy;
import org.apache.wicket.model.IModel;
import org.apache.wicket.protocol.http.WebApplication;
import org.apache.wicket.request.Request;
import org.apache.wicket.request.Response;
import org.apache.wicket.request.cycle.RequestCycle;
import org.apache.wicket.request.http.WebRequest;
import org.apache.wicket.request.http.WebResponse;
import org.apache.wicket.request.mapper.parameter.PageParameters;
import org.apache.wicket.response.StringResponse;
import org.apache.wicket.util.string.Strings;
import org.apache.wicket.util.visit.IVisit;
import org.apache.wicket.util.visit.IVisitor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * Base class for HTML pages. This subclass of Page simply returns HTML when asked for its markup
 * type. It also has a method which subclasses can use to retrieve a bookmarkable link to the
 * application's home page.
 * <p>
 * WebPages can be constructed with any constructor when they are being used in a Wicket session,
 * but if you wish to link to a Page using a URL that is "bookmarkable" (which implies that the URL
 * will not have any session information encoded in it, and that you can call this page directly
 * without having a session first directly from your browser), you need to implement your Page with
 * a no-arg constructor or with a constructor that accepts a PageParameters argument (which wraps
 * any query string parameters for a request). In case the page has both constructors, the
 * constructor with PageParameters will be used.
 * 
 * @author Jonathan Locke
 * @author Eelco Hillenius
 * @author Juergen Donnerstag
 * @author Gwyn Evans
 */
public class WebPage extends Page
{
	/** log. */
	private static final Logger log = LoggerFactory.getLogger(WebPage.class);

	private static final long serialVersionUID = 1L;

	/**
	 * Constructor.
	 *
	 * Having this constructor public means that your page is 'bookmarkable' and hence
	 * can be called/ created from anywhere.
	 */
	protected WebPage()
	{
		commonInit();
	}

	/**
	 * @see Page#Page(IModel)
	 */
	protected WebPage(final IModel<?> model)
	{
		super(model);
		commonInit();
	}

	/**
	 * Constructor which receives wrapped query string parameters for a request. Having this
	 * constructor public means that your page is 'bookmarkable' and hence can be called/ created
	 * from anywhere. For bookmarkable pages (as opposed to when you construct page instances
	 * yourself, this constructor will be used in preference to a no-arg constructor, if both exist.
	 * Note that nothing is done with the page parameters argument. This constructor is provided so
	 * that tools such as IDEs will include it their list of suggested constructors for derived
	 * classes.
	 * 
	 * Please call this constructor if you want to remember the pageparameters
	 * {@link #getPageParameters()}. So that they are reused for stateless links.
	 * 
	 * @param parameters
	 *            Wrapped query string parameters.
	 */
	protected WebPage(final PageParameters parameters)
	{
		super(parameters);
		commonInit();
	}

	/**
	 * Gets the markup type for a WebPage, which is "html" by default. Support for pages in another
	 * markup language, such as VXML, would require the creation of a different Page subclass in an
	 * appropriate package under org.apache.wicket.markup. To support VXML (voice markup), one might
	 * create the package org.apache.wicket.markup.vxml and a subclass of Page called VoicePage.
	 * 
	 * @return Markup type for HTML
	 */
	@Override
	public MarkupType getMarkupType()
	{
		return MarkupType.HTML_MARKUP_TYPE;
	}

	/**
	 * Common code executed by constructors.
	 */
	private void commonInit()
	{
		// so far a noop
	}

	@Override
	protected void onRender()
	{
		// Configure the response such as headers etc.
		configureResponse((WebResponse)RequestCycle.get().getResponse());

		// The rules if and when to insert an xml decl in the response are a bit tricky. Allow the
		// user to replace the default per page and per application.
		renderXmlDecl();

		super.onRender();
	}

	/**
	 * The rules if and when to insert an xml decl in the response are a bit tricky. Allow the user
	 * to replace the default per page and per application.
	 */
	protected void renderXmlDecl()
	{
		WebApplication.get().renderXmlDecl(this, false);
	}

	/**
	 * Set-up response with appropriate content type, locale and encoding. The locale is set equal
	 * to the session's locale. The content type header contains information about the markup type
	 * (@see #getMarkupType()) and the encoding. The response (and request) encoding is determined
	 * by an application setting (@see ApplicationSettings#getResponseRequestEncoding()). If null,
	 * no xml decl will be written.
	 * 
	 * @param response
	 *            The WebResponse object
	 */
	protected void configureResponse(final WebResponse response)
	{
		// Users may subclass setHeader() to set there own headers
		setHeaders(response);

		// The response encoding is an application setting
		final String encoding = getApplication().getRequestCycleSettings()
			.getResponseRequestEncoding();
		final boolean validEncoding = (Strings.isEmpty(encoding) == false);
		final String contentType;
		if (validEncoding)
		{
			contentType = getMarkupType().getMimeType() + "; charset=" + encoding;
		}
		else
		{
			contentType = getMarkupType().getMimeType();
		}

		response.setContentType(contentType);
	}

	/**
	 * Subclasses can override this to set there headers when the Page is being served. By default
	 * these headers are set:
	 * 
	 * <pre>
	 * response.setHeader(&quot;Date&quot;, &quot;[now]&quot;);
	 * response.setHeader(&quot;Expires&quot;, &quot;[0]&quot;);
	 * response.setHeader(&quot;Pragma&quot;, &quot;no-cache&quot;);
	 * response.setHeader(&quot;Cache-Control&quot;, &quot;no-cache&quot;);
	 * </pre>
	 * 
	 * So if a Page wants to control this or doesn't want to set this info it should override this
	 * method and don't call super.
	 * 
	 * @param response
	 *            The WebResponse where set(Date)Header can be called on.
	 */
	protected void setHeaders(WebResponse response)
	{
		response.disableCaching();
	}

	/**
	 * 
	 * @see org.apache.wicket.Component#onAfterRender()
	 */
	@Override
	protected void onAfterRender()
	{
		// only in development mode validate the headers
		if (getApplication().usesDevelopmentConfig())
		{
			// check headers only when page was completely rendered
			if (wasRendered(this))
			{
				validateHeaders();
			}
		}

		super.onAfterRender();
	}

	/**
	 * Validate that each component which wanted to contribute to the header section actually was
	 * able to do so.
	 */
	private void validateHeaders()
	{
		// search for HtmlHeaderContainer in the first level of children or deeper
		// if there are transparent resolvers used
		HtmlHeaderContainer header = visitChildren(new IVisitor<Component, HtmlHeaderContainer>()
		{
			@Override
			public void component(final Component component, final IVisit<HtmlHeaderContainer> visit)
			{
				if (component instanceof HtmlHeaderContainer)
				{
					visit.stop((HtmlHeaderContainer)component);
				}
				else if (component instanceof TransparentWebMarkupContainer == false)
				{
					visit.dontGoDeeper();
				}
			}
		});

		if (header == null)
		{
			// the markup must at least contain a <body> tag for wicket to automatically
			// create a HtmlHeaderContainer. Log an error if no header container
			// was created but any of the components or behaviors want to contribute
			// something to the header.
			header = new HtmlHeaderContainer(HtmlHeaderSectionHandler.HEADER_ID);
			add(header);

			RequestCycle requestCycle = getRequestCycle();
			Response orgResponse = requestCycle.getResponse();
			try
			{
				StringResponse tempResponse = new StringResponse();
				requestCycle.setResponse(tempResponse);

				// Render all header sections of all components on the page
				AbstractHeaderRenderStrategy.get().renderHeader(header, null, getPage());

				IHeaderResponse headerResponse = header.getHeaderResponse();
				headerResponse.close();
				CharSequence collectedHeaderOutput = tempResponse.getBuffer();
				if (collectedHeaderOutput.length() > 0)
				{
					reportMissingHead(collectedHeaderOutput);
				}
			}
			catch (Exception e)
			{
				// just swallow this exception, there isn't much we can do about.
				log.error("header/body check throws exception", e);
			}
			finally
			{
				this.remove(header);
				requestCycle.setResponse(orgResponse);
			}
		}
	}

	/**
	 * Reports an error that there is no &lt;head&gt; and/or &lt;body&gt; in the page and
	 * there is no where to write the header response.
	 * 
	 * @param collectedHeaderOutput
	 *          The collected response that should have been written to the &lt;head&gt;
	 * @see #validateHeaders()
	 */
	protected void reportMissingHead(final CharSequence collectedHeaderOutput)
	{
		log.error("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^");
		log.error("You probably forgot to add a <body> or <head> tag to your markup since no Header Container was \n" +
				"found but components were found which want to write to the <head> section.\n" +
				collectedHeaderOutput);
		log.error("^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^");
	}

	/**
	 * Creates and returns a bookmarkable link to this application's home page.
	 * 
	 * @param id
	 *            Name of link
	 * @return Link to home page for this application
	 */
	protected final BookmarkablePageLink<Void> homePageLink(final String id)
	{
		return new BookmarkablePageLink<>(id, getApplication().getHomePage());
	}

	/**
	 * Prevents page from get dirty inside an AJAX request.
	 */
	@Override
	public final void dirty(boolean isInitialization)
	{
		Request request = getRequest();
		if (isInitialization == false && request instanceof WebRequest &&
			((WebRequest)request).isAjax())
		{
			return;
		}
		super.dirty(isInitialization);
	}
}
