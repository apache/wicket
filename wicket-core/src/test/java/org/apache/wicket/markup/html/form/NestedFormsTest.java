/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.wicket.markup.html.form;

import java.util.ArrayList;
import java.util.List;

import org.apache.wicket.Component;
import org.apache.wicket.MarkupContainer;
import org.apache.wicket.WicketTestCase;
import org.apache.wicket.markup.IMarkupResourceStreamProvider;
import org.apache.wicket.markup.html.WebPage;
import org.apache.wicket.util.resource.IResourceStream;
import org.apache.wicket.util.resource.StringResourceStream;
import org.apache.wicket.util.tester.FormTester;
import org.junit.Test;

/**
 * @author Pedro Santos
 */
public class NestedFormsTest extends WicketTestCase
{
	/**
	 *
	 */
	@Test
	public void postOrderSequenceSubmittingRootForm()
	{
		tester.startPage(TestPage.class);
		TestPage testPage = (TestPage)tester.getLastRenderedPage();
		FormTester formTester = tester.newFormTester("outerForm");
		formTester.submit("outerSubmit");
		assertEquals(3, testPage.submitSequence.size());
		assertEquals(0, testPage.submitSequence.indexOf(testPage.outerSubmit));
		assertEquals(1, testPage.submitSequence.indexOf(testPage.innerForm));
		assertEquals(2, testPage.submitSequence.indexOf(testPage.outerForm));
	}

	/**
	 *
	 */
	@Test
	public void postOrderSequenceSubmittingInnerForm()
	{
		tester.startPage(TestPage.class);
		TestPage testPage = (TestPage)tester.getLastRenderedPage();
		FormTester formTester = tester.newFormTester("outerForm");
		formTester.submit("innerForm:innerSubmit");
		assertEquals(2, testPage.submitSequence.size());
		assertEquals(0, testPage.submitSequence.indexOf(testPage.innerSubmit));
		assertEquals(1, testPage.submitSequence.indexOf(testPage.innerForm));
	}

	/** */
	public static class TestPage extends WebPage implements IMarkupResourceStreamProvider
	{
		/** */
		private static final long serialVersionUID = 1L;
		private List<Component> submitSequence = new ArrayList<Component>();
		Form<Void> outerForm;
		Button outerSubmit;
		Form<Void> innerForm;
		Button innerSubmit;

		/**
		 * Construct.
		 */
		public TestPage()
		{
			outerForm = new Form<Void>("outerForm")
			{
				/** */
				private static final long serialVersionUID = 1L;

				@Override
				protected void onSubmit()
				{
					submitSequence.add(this);
				}
			};
			add(outerForm);
			outerSubmit = new Button("outerSubmit")
			{
				/** */
				private static final long serialVersionUID = 1L;

				@Override
				public void onSubmitBeforeForm()
				{
					submitSequence.add(this);
				}
			};
			outerForm.add(outerSubmit);
			innerForm = new Form<Void>("innerForm")
			{
				/** */
				private static final long serialVersionUID = 1L;

				@Override
				protected void onSubmit()
				{
					submitSequence.add(this);
				}
			};
			outerForm.add(innerForm);
			innerSubmit = new Button("innerSubmit")
			{
				/** */
				private static final long serialVersionUID = 1L;

				@Override
				public void onSubmitBeforeForm()
				{
					submitSequence.add(this);
				}
			};
			innerForm.add(innerSubmit);
		}

		@Override
		public IResourceStream getMarkupResourceStream(MarkupContainer container,
			Class<?> containerClass)
		{
			return new StringResourceStream(
				"<html><body>"
					+ "<form wicket:id=\"outerForm\">"//
					+ "  <input type=\"submit\" wicket:id=\"outerSubmit\"/>"//
					+ "  <form wicket:id=\"innerForm\"><input type=\"submit\" wicket:id=\"innerSubmit\"/></form>"//
					+ "</form>" + //
					"</body></html>");
		}
	}
}
