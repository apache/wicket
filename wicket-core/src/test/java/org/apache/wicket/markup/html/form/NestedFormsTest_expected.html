<html><body><form wicket:id="outerForm" id="outerForm1" method="post" action="./page?0-1.-outerForm"><div id="outerForm1_hf_0" style="width:0px;height:0px;position:absolute;left:-100px;top:-100px;overflow:hidden" class="hidden-fields"><input type="hidden" name="0-1.-outerForm" value="" /></div>  <input type="submit" wicket:id="outerSubmit" name="outerSubmit" id="outerSubmit2"/>  <div wicket:id="innerForm" id="innerForm3"><input type="submit" wicket:id="innerSubmit" name="innerForm:innerSubmit" id="innerSubmit4"/></div></form></body></html>